### PR TITLE
[pythonic config] Add support for Pythonic config classes to @configured

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Adds support for taking in and returning Pythonic config classes with the `@configured` decorator. This is useful for wrapping ops with lots of config options.

```python

class MyComplexOpConfig(Config):
  a_bool: bool
  an_int: int
  a_string: str


@op
def my_op(config: MyComplexOpConfig):
   ...


class MySimplifiedOpConfig(Config):
  an_int: int

@configured(my_op)
def my_simpler_op(config: MySimplifiedOpConfig):
  return MyComplexOpConfig(a_bool=True, a_string="foo", an_int=config.an_int)
```

## Test Plan

New unit tests, more forthcoming

### CODE REVIEW

**Comment by benpankow** [comment]
Current dependencies on/for this PR:
* master
  * **PR #14624** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/14624" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/14624?utm_source=stack-comment).

**Comment by schrockn** [comment]
How did this come up? User feedback?

**Comment by benpankow** [comment]
> How did this come up? User feedback?

Yes, a user was interested in a pattern where they write a generic op with many config options and then construct a variety of `@configured` ops for use in jobs

File: python_modules/dagster/dagster/_core/definitions/configurable.py [hunk 1]
```diff
@@ -192,6 +192,64 @@ def _check_configurable_param(configurable: ConfigurableDefinition) -> None:
...
+
+    config_fn_params = get_function_params(user_fn)
+    check.invariant(len(config_fn_params) == 1, "Config mapping should have exactly one parameter")
+
+    param = config_fn_params[0]
+
+    # If the parameter is a subclass of Config, we can infer the config schema from the
+    # type annotation. We'll also wrap the config mapping function to convert the config
+    # dictionary into the appropriate Config object.
+    if safe_is_subclass(param.annotation, Config):
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/configurable.py:225]: prefer early return here too.

**Review by schrockn — CHANGES_REQUESTED** [review]
Looks good overall. Let's update appropriate docblocks/docs in the same PR (or in a stack).

**Review by schrockn — APPROVED** [review]
Great 👍🏻

**Pull request merged**